testsys: `aws-k8s` migration testing with `cargo make test` (#2273)
bottlerocket-os/bottlerocket: TESTING.md

  test
```

```shell
cargo make watch-test
```

## Migration Testing

Migration testing starts testsys at a starting version, or a provided initial image id and migrates instances to the target version.
In order to accomplish this a few artifacts need to be created:
* `TUF` repos that are publicly accessible

> * A previous release of bottlerocket signed with available keys

Reviewer: Nit: `sed 's/bottlerocket/Bottlerocket'`

> * The ami id from the previous release

Reviewer: Nit: `sed 's/ami/AMI'`

> * Build images and repos for current changes

Reviewer: Nit: a link to the documentation that explains how to build and publish repos would be nice

### The setup

#### Prepare `Infra.toml`

> A publicly available s3 bucket is needed to complete the next steps, once a bucket is created the steps may be continued.

Reviewer: This is not true, a contributor can use whatever they want to store their TUF repositories, so it will be best to say "the URL of a TUF repository" or something along those lines.

Author: The example steps I provide include syncing tuf repos with the pre-update metadata. I was trying to make the process of onboarding to testsys as easy as possible by providing a sample workflow from start to finish. I can add to the beginning that it is not required to have, but this example will use it.

Reviewer: I think it will be best to point whomever reads this guide to [this document](https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location) so that they are aware that they are free to decide where to store their TUF repo. I would definitely remove the "publicly" available part since that isn't a good practice, and we shouldn't advocate for that. It will be best to say "use the URL of an accessible TUF repo"; and for your example, assume the contributor followed the link you will add, and that they have a URL that is on top of a properly configured S3 bucket (that's what ultimately happened in the document I linked).

> `Infra.toml` is used by testsys to determine tuf repo locations, so `metadata_base_url` and `targets_base_url` need to be set based on the s3 bucket. The examples below assume `metadata_base_url` is the top level directory of the bucket and `targets_base_url` is a directory named `targets` in the top level directory. 

Reviewer: Nit: a line per sentence

> The examples below also assume that the default repo in `Infra.toml`, but any repo can be used by setting the `PUBLISH_REPO` environment variable.

Reviewer: I think there is something missing in this phrase?

```suggestion
The examples below also assume that the default repo is used in `Infra.toml`, but any repo can be used by setting the `PUBLISH_REPO` environment variable.
```

#### Starting bottlerocket images

In this example we will use `v1.8.0` as our starting bottlerocket version, but any tag from bottlerocket will work. 

> The following bash will checkout the proper branch from git and create the build images and repos for testing.

Reviewer: ```suggestion
The following bash script will checkout the proper branch from git and create the build images and TUF repos for testing.
```

```shell
export TESTSYS_STARTING_VERSION="1.8.0"

> git checkout v${TESTSYS_STARTING_VERSION}

Reviewer: ```suggestion
git checkout "v${TESTSYS_STARTING_VERSION}"
```

cargo make
cargo make ami
cargo make repo
```

> Next, the repos need to be added to the `TUF` repos (s3 bucket). Make sure to change `S3_BUCKET_NAME` to the `TUF` repo bucket name.

Reviewer: I think all mentions of a S3 bucket should be removed, folks in the community may have their own way to update their TUF repos.

Author: See above comment. Should we remove the step by step guide so that we can remove the s3 bucket bits?

```shell
S3_BUCKET_NAME=<s3 bucket name>
aws s3 sync build/repos/default/latest s3://${S3_BUCKET_NAME}/
```

The last step with the older bottlerocket version is retrieving the ami that was created.

```shell
BUILDSYS_ARCH="x86_64"
BUILDSYS_VARIANT="aws-k8s-1.21"
REGION="us-west-2"

AMIS_JSON="build/images/${BUILDSYS_ARCH}-${BUILDSYS_VARIANT}/latest/bottlerocket-${BUILDSYS_VARIANT}-${BUILDSYS_ARCH}-amis.json"
export TESTSYS_STARTING_IMAGE_ID=$(jq -r --arg r "${REGION}" ' .[$r].id' "${AMIS_JSON}")
echo "Starting ami for migration tests '${TESTSYS_STARTING_IMAGE_ID}'"
```

Make note of the ami that is printed, it will be needed in future steps.

#### Target bottlerocket images

Now, it's time to create the bottlerocket artifacts that need to be upgraded to.

Switch to the working git branch that should be built from.

```shell
WORKING_BRANCH="develop"

> git checkout ${WORKING_BRANCH}

Reviewer: Nit:

```suggestion
git checkout "${WORKING_BRANCH}"
```

```

> Now, [Release.toml](Release.toml) may need to be updated.

Reviewer: For this testing it must, right?

Author: For using develop yes. But if someone wanted to try building 1.6.0 and 1.6.1 to see how testsys migrations work they wouldn't need to.

>   test

Reviewer: I think the two commits can be squashed in one, since they are both related to the same change.

```

```shell
cargo make watch-test
```

> ## Migration Testing

Reviewer 2: Some clarification on what "migration" means in this context would be nice, i.e. to update instances and downgrade them.

Migration testing starts testsys at a starting version, or a provided initial image id and migrates

> instances to the target version.

Reviewer 3: I think typically we let the markdown renderer handle the line-wrapping.

Reviewer 2: nit: slightly different wording to hopefully make it more clear
```suggestion
Migration testing launches instances of a starting Bottlerocket version, or a provided initial AMI and migrates instances to the target version.
```

In order to accomplish this a few artifacts need to be created:

> * Tuf repos that are publicly accessible

Reviewer 3: nit. `TUF` repos

* A previous release of bottlerocket signed with available keys

> * The ami id from the previous release

Reviewer 2: nit:
```suggestion
* The AMI ID of the previous release
```

> * Build images and repos for current changes

Reviewer 2: nit: more clarification
```suggestion
* Image artifacts and local TUF repos of said artifacts for current changes
```

### The setup

#### Prepare `Infra.toml`

A publicly available s3 bucket is needed to complete the next steps, once a bucket is created the

> steps may be continued.

Reviewer 2: nit:
```suggestion
A publicly available S3 bucket is needed to complete the next steps.
```

> Infra.toml is used by testsys to determine tuf repo locations, so `metadata_base_url` and

Reviewer 3: nit. `Infra.toml`

Reviewer 2: nit: capitalize TUF and S3

* A publicly accessible TUF repository
* A previous release of Bottlerocket signed with available keys
* The AMI ID for the previous release
* Image artifacts and local TUF repos of said artifacts for current changes

### The setup

#### Prepare `Infra.toml`

We need the URL of an accessible TUF repo so the Bottlerocket instances know where to retrieve the update metadata and targets. 
Follow our (publishing guide)[PUBLISHING.md#repo-location] to set up TUF repos.
`Infra.toml` is used by testsys to determine TUF repo locations, so `metadata_base_url` and `targets_base_url` need to be set based on the repo that was just created. 
The examples below also assume that the default repo is being used in `Infra.toml`, but any repo can be used by setting the `PUBLISH_REPO` environment variable.

#### Starting Bottlerocket images

In this example we will use `v1.9.0` as our starting Bottlerocket version, but any tag from Bottlerocket will work. 
The following bash script will checkout the proper branch from git and create the build images and TUF repos for testing.

```shell
git checkout "v1.9.0"
cargo make
cargo make ami
cargo make repo
```

> Next, the local repos need to be added to the TUF repos. 

Reviewer 4: ```suggestion
Remember to sync your TUF repos with the new metadata and targets.
```

The last step with the older Bottlerocket version is retrieving the AMI that was created.

```shell
BUILDSYS_ARCH="x86_64"
BUILDSYS_VARIANT="aws-k8s-1.21"
REGION="us-west-2"

> AMIS_JSON="build/images/${BUILDSYS_ARCH}-${BUILDSYS_VARIANT}/latest/bottlerocket-${BUILDSYS_VARIANT}-${BUILDSYS_ARCH}-amis.json"

Reviewer 4: Is this step still needed for something?

  test
```

```shell
cargo make watch-test
```

## Migration Testing

Migration testing starts testsys at a starting version, or a provided initial image id and migrates instances to the target version.
In order to accomplish this a few artifacts need to be created:

> * A `TUF` repo

Reviewer 2: ```suggestion
* A publicly accessible TUF repository
```

* A previous release of Bottlerocket signed with available keys

> * The AMI id from the previous release

Reviewer 2: nit: 
```suggestion
* The AMI ID for the previous release
```

  test
```

```shell
cargo make watch-test
```

## Migration Testing

Migration testing is used to ensure Bottlerocket can update from one version to a new version and back. This involves launching Bottlerocket instances, upgrading them, and downgrading them.

Migration testing launches instances of a starting Bottlerocket version, or a provided initial AMI and migrates instances to the target version.
In order to accomplish this a few artifacts need to be created:
* A publicly accessible TUF repository
* A previous release of Bottlerocket signed with available keys
* The AMI ID for the previous release

> * Image artifacts and local TUF repos of said artifacts for current changes

Reviewer 4: This is a reasonable set of requirements, but that there's also a clear need for a tool to help meet the requirements. It's tricky enough to accomplish that ten developers will find ten different ways of doing it.

I'm not sure about taking an AMI ID. That wouldn't apply to either VMware or bare metal. I'd prefer to understand what the upgrade/downgrade solution looks like on other platforms first.

If you weren't allowed to use either AMIs or SSM parameters, how would you approach this? Can it be done with just the contents of the repo?

Reviewer 4: The current implementation addresses my concern here, thanks!

By automating the lookup of the previous version and the AMI, it's more likely that the developer will follow the expected steps rather than going "off script" to force the use of AMIs built some other way.

### The setup

#### Prepare `Infra.toml`

> We need the URL of an accessible TUF repo so the Bottlerocket instances know where to retrive the update metadata and targets. 

Reviewer 4: I spotted this typo, might it be good to run the guide through a spell checker:

```suggestion
We need the URL of an accessible TUF repo so the Bottlerocket instances know where to retrieve the update metadata and targets. 
```

Follow our (publishing guide)[PUBLISING.md#repo-location] to set up TUF repos.
`Infra.toml` is used by testsys to determine TUF repo locations, so `metadata_base_url` and `targets_base_url` need to be set based on the repo that was just created. 
The examples below also assume that the default repo is being used in `Infra.toml`, but any repo can be used by setting the `PUBLISH_REPO` environment variable.

#### Starting Bottlerocket images

In this example we will use `v1.8.0` as our starting Bottlerocket version, but any tag from Bottlerocket will work. 
The following bash script will checkout the proper branch from git and create the build images and TUF repos for testing.

```shell
export TESTSYS_STARTING_VERSION="1.8.0"

> git checkout "v${TESTSYS_STARTING_VERSION}"

Reviewer 4: We should be consistent about whether or not a `v` prefix is required, and possibly support both forms in environment variables like `TESTSYS_STARTING_VERSION`.

cargo make
cargo make ami
cargo make repo
```

Next, the local repos need to be added to the TUF repos. 

The last step with the older Bottlerocket version is retrieving the AMI that was created.

```shell
BUILDSYS_ARCH="x86_64"
BUILDSYS_VARIANT="aws-k8s-1.21"
REGION="us-west-2"

AMIS_JSON="build/images/${BUILDSYS_ARCH}-${BUILDSYS_VARIANT}/latest/bottlerocket-${BUILDSYS_VARIANT}-${BUILDSYS_ARCH}-amis.json"
export TESTSYS_STARTING_IMAGE_ID=$(jq -r --arg r "${REGION}" ' .[$r].id' "${AMIS_JSON}")
echo "Starting AMI for migration tests '${TESTSYS_STARTING_IMAGE_ID}'"
```

Make note of the AMI that is printed, it will be needed in future steps.

#### Target Bottlerocket images

Now, it's time to create the Bottlerocket artifacts that need to be upgraded to.

Switch to the working git branch that should be built from.

```shell
WORKING_BRANCH="develop"
git checkout "${WORKING_BRANCH}"
```

Now, [Release.toml](Release.toml) may need to be updated.
The version at the top of the file should be changed to whatever version migrations should be tested for.

> In the example using `v1.8.0`, the version should be changed to `1.9.0`.

Reviewer 4: This is one of the places where "`v` or not `v`" is awkward.
